hoot-core/src/main/cpp/hoot/core/io/OsmJsonReader.cpp

```diff
@@ -0,0 +1,410 @@
+  _useDataSourceIds(true),
+  _defaultCircErr(15.0),
+  _propTree(),
+  _version(""),
+  _generator(""),
+  _timestamp_base(""),
+  _copyright(""),
+  _url(""),
+  _isFile(false),
+  _isWeb(false)
+{
+  // Do nothing special
+}
+
+OsmJsonReader::~OsmJsonReader()
+{
+  close();
+}
+
+bool OsmJsonReader::isSupported(QString url)
+{
+  QUrl myUrl(url);
+
+  // Is it a file?
+  if (myUrl.isLocalFile())
+  {
+    QString filename = myUrl.toLocalFile();
+
+    if (QFile::exists(filename))
+      return true;
+  }
+
+  // Is it a web address?
+  if ("http" == myUrl.scheme() || "https" == myUrl.scheme())
+  {
+    return true;
+  }
+
+  // Default to not supported
+  return false;
+}
+
+/**
+ * Opens the specified URL for reading.
+ */
+void OsmJsonReader::open(QString url)
+{
+  try
+  {
+    // Bail out if unsupported
+    if (!isSupported(url))
+      return;
+
+    // Handle files or URLs
+    _url = QUrl(url);
+    if (_url.isLocalFile())
+    {
+      _isFile = true;
+      _file.setFileName(_url.toLocalFile());
+      _file.open(QFile::ReadOnly | QFile::Text);
+    }
+    else
+    {
+      _isWeb = true;
+    }
+  }
+  catch (std::exception ex)
+  {
+    ostringstream oss;
+    oss << "Exception opening URL (" << url << "): " << ex.what();
+    throw HootException(QString::fromStdString(oss.str()));
+  }
+}
+
+void OsmJsonReader::close()
+{
+  if (_isFile)
+    _file.close();
+}
+
+/**
+ * Reads the specified map. When this method is complete
+ * the input will likely be closed.
+ */
+void OsmJsonReader::read(shared_ptr<OsmMap> map)
+{
+  if (_isFile)
+  {
+    QTextStream instream(&_file);
+    QString jsonStr = instream.readAll();
+    _loadJSON(jsonStr);
+    _parseOverpassJson(map);
+  }
+  else
+  {
+    boost::shared_ptr<QNetworkAccessManager> pNAM(new QNetworkAccessManager());
+    QNetworkRequest request(_url);
+    boost::shared_ptr<QNetworkReply> pReply(pNAM->get(request));
+
+    // Wait for finished signal from reply object
+    QEventLoop loop;
+    QObject::connect(pReply.get(), SIGNAL(finished()), &loop, SLOT(quit()));
+    loop.exec();
+
+    QByteArray data = pReply->readAll();
+    QString jsonStr = QString::fromAscii(data.data());
+    _loadJSON(jsonStr);
+    _parseOverpassJson(map);
+  }
+}
+
+// Throws HootException on error
+void OsmJsonReader::_loadJSON(QString jsonStr)
+{
+  // Clear out anything that might be hanging around
+  _propTree.clear();
+
+  // Handle single or double quotes
+  _scrubQuotes(jsonStr);
+
+  // Convert string to stringstream
+  stringstream ss(jsonStr.toUtf8().constData(), ios::in);
+
+  if (!ss.good())
+  {
+    throw HootException(QString("Error reading from JSON string:\n%1").arg(jsonStr));
+  }
+
+  try
+  {
+    pt::read_json(ss, _propTree);
+  }
+  catch (pt::json_parser::json_parser_error e)
+  {
+    QString reason = QString::fromStdString(e.message());
+    QString line = QString::number(e.line());
+    throw HootException("Error parsing JSON " + reason + "(line " + line + ")");
```

Comment:
There is nothing wrong with this, but you can also use QString's arg methods. E.g.

```
throw HootException(QString("Error parsing JSON %1 (line %2)").arg(e.message()).arg(e.line()));
```

FWIW